modules/nf-core/fq/lint/main.nf

```diff
@@ -29,5 +30,10 @@ process FQ_LINT {
     "${task.process}":
         fq: \$(echo \$(fq lint --version | sed 's/fq-lint //g'))
     END_VERSIONS
+
+    if ! tail -n 1 ${prefix}.fq_lint.txt | grep -q 'fq-lint end'; then
```

Comment:
We shouldn't change the module, generally we try to let the tools work as they do, without overlaying additional logic.

We can add some logic to the subworkflow + workflow that would filter out any libraries failing linting. You may already see logic in rnaseq that we use for trimming, strand failures etc, and we can use the same mechanism.

But my understanding is that fq_lint will return a non-zero error code on a failure? That should trigger a stop without this. Am I mistaken?

Reply:
@pinin4fjords Yeah, kind of. fq lint has different validators that are all assigned a [different code](https://github.com/stjude-rust-labs/fq#lint), and if a FastQ file fails the linting, you will see in the log what code is failed on. You can't grep for the codes themselves in the log, as the log will say upfront what validators are enabled by code. That's why I had to grep for "fq-lint end" and check if it **_doesn't_** exist. That's the list line you'll see in the linting log if the linting was successful, so failed FastQ files won't have that in their log.

So, fq lint doesn't actually put out an error code itself. If we left the current module alone as is, then that process will successfully complete on any FastQ file, even corrupt ones, which defeats the purpose of adding the linting into the pipeline.

I am definitely open to trying out some different logic to filter out samples that failed linting. Are you thinking those samples would just filtered out, and the pipeline would keep going with the good samples, or that the pipeline would exit if it finds a sample that failed linting? We find that our users often want the pipeline to stop so that they can try reuploading (or redownloading from source and then reuploading) the offending FastQ file, which has worked in a lot of cases. Maybe another conditional that could be set as to if a user wants the pipeline to exit if a sample fails linting or if they want the pipeline to continue with successful samples?

Reply:
What I mean is the following. Assume a bad fastq file, foo.fq, with content like:

```
@SEQ_ID_1
GATTTGGGGTTTTCCCAGTT
THIS SHOULD NOT BE HERE
IIIIIIIIIIIIIIIIIIII
@SEQ_ID_2
AGCTAGCTAGGCTAGCTAAG
+
JJJJJJJJJJJJJJJJJJJJ
```

The output of `fq lint foo.fq` is: 

```
2024-11-17T17:03:29.770623Z  INFO fq::commands::lint: fq-lint start
2024-11-17T17:03:29.770645Z  INFO fq::commands::lint: validating single end read
2024-11-17T17:03:29.770651Z  INFO fq::validators: disabled validators: []
2024-11-17T17:03:29.770658Z  INFO fq::validators: enabled single read validators: ["[S003] NameValidator", "[S004] CompleteValidator", "[S002] AlphabetValidator", "[S001] PlusLineValidator", "[S005] ConsistentSeqQualValidator", "[S006] QualityStringValidator"]
2024-11-17T17:03:29.770668Z  INFO fq::validators: enabled paired read validators: []
2024-11-17T17:03:29.770671Z  INFO fq::commands::lint: starting validation
/data/foo.fq:3:1: [S001] PlusLineValidator: missing + prefix
```

This produces a non-zero error code:

```
$ echo $?
1
```

Because Nextflow runs processes with `set -e` turned on, the task will fail when it encounters that issue. 

I see that `fq` has its own codes, but that wasn't what I meant.

Reply:
For the specific rnaseq example, we'd just make sure to set the appropriate label so the process gets enough resource to start. 

But to the general point, you can use dynamic retry strategies to catch exit codes reflecting e.g. OOM:

```
errorStrategy { task.exitStatus in 137..140 ? 'retry' : 'terminate' }
```

Reply:
Actually you could PR against [this](https://github.com/nf-core/modules/blob/af84433c2e619cf28e026d4140b1b8763ae3690b/modules/nf-core/fq/lint/main.nf#L3), if you feel the current value isn't appropriate.